lib/puppet/configurer.rb

```diff
@@ -181,25 +181,50 @@ def apply_catalog(catalog, options)
   # which accepts :tags and :ignoreschedules.
   def run(options = {})
     pool = Puppet::Network::HTTP::Pool.new(Puppet[:http_keepalive_timeout])
+    # We create the report pre-populated with default settings for
+    # environment and transaction_uuid very early, this is to ensure
+    # they are sent regardless of any catalog compilation failures or
+    # exceptions.
+    options[:report] ||= Puppet::Transaction::Report.new("apply", nil, @environment, @transaction_uuid)
+    report = options[:report]
```

Comment:
Report initialization is moved up here so that logging around master failover ends up in the report (and also for PUP-6378 adding the selected master to the report)